daemon/daemon_linux.go

```diff
@@ -14,13 +14,33 @@ import (
 	"github.com/docker/docker/pkg/mount"
 	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
+	"golang.org/x/sys/unix"
 )
 
 const (
 	defaultResolvConf   = "/etc/resolv.conf"
 	alternateResolvConf = "/run/systemd/resolve/resolv.conf"
 )
 
+// On Linux, we need to ensure that parallelStartupJobs doesn't cause us to
+// exceed RLIMIT_NOFILE. If parallelStartupJobs is too large, we reduce it and
+// give a warning (since in theory the user should increase their ulimits to
+// the largest possible value for dockerd).
+func init() {
```

Comment:
Is this neccessary? In reality we don't know how many containers are running here and seems a bit heavy...

If we do keep this, I think we should call it during daemon init instead of package init.

Reply:
We could also take into account the number of containers, and warn that the limit is low while also adjusting the semaphore size accordingly.

Reply:
I would suggest running Docker with an `NOFILE` limit of `1024` (the systemd default) will certainly cause many other unrelated problems.

But sure, I can make it happen depending on how many containers we actually have.

Reply:
Sorry, another problem with adding it to `daemon/daemon.go` is that Windows doesn't have `ulimit`s -- which is why I've only made the override happen in `daemon_linux.go`.

Reply:
I will also switch to `golang.org/x/sync/semaphore`.